specification/devcenter/data-plane/Microsoft.DevCenter/preview/2023-07-01-preview/environments.json

```diff
@@ -944,6 +1110,79 @@
           }
         ]
       }
+    },
+    "EnvironmentOperationListResult": {
+      "type": "object",
+      "description": "The operations list result.",
+      "properties": {
+        "value": {
+          "description": "The current page of results.",
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/EnvironmentOperation"
+          }
+        },
+        "nextLink": {
+          "type": "string",
+          "description": "The URL to get the next set of results."
+        }
+      },
+      "required": [
+        "value"
+      ]
+    },
+    "EnvironmentOperation": {
+      "type": "object",
+      "description": "An operation on an environment.",
+      "properties": {
+        "id": {
+          "description": "Unique identifier for the environment operation.",
+          "type": "string",
+          "readOnly": true
```

Comment:
We recommend against `readOnly` properties. Marking even response properties as `required` (the `required` array, rather) is better, but also atypical for response models - especially for intrinsic (non-pointer/references e.g., `int`), since this generates e.g., in C# `int` vs `int?` and changing it later is a breaking change. Better that they are all possibly null if you make changes later.

Reply:
Updated this to remove `readOnly` on the new types, and confirmed that now we only have `readOnly` on models that are sent to the service by users (Dev Box and Environment models). We currently have `required` for a minimal property set (for operations, `operationId`, `uri`, `status`, and the `kind` discriminator), which we would expect users to rely on always being present. We can definitely remove that for new models if the guidance is to avoid using it at all now, we marked the minimal set as required to avoid nullability on very basic properties users can rely on being present.